### PR TITLE
MODULES-10763 Only report apt-get update as a corrective change if /var/cache/apt/pkgcache.bin changes.

### PR DESCRIPTION
**Pull request by pillarsdotnet**
Instead of running `apt-get update` as an exec directly, we run it as the `unless` attribute of an exec.

After running `apt-get update`, if the content of `/var/cache/apt/pkgcache.bin` is unchanged, the exec does not run, and reports no change.

If the content of `/var/cache/apt/pkgcache.bin` changes, the exec runs and reports a corrective change.

This resolves [MODULES-10763](https://tickets.puppetlabs.com/browse/MODULES-10763).

### CODE REVIEW

File: manifests/update.pp [hunk 1]
```diff
@@ -56,13 +56,20 @@
   } else {
     $_refresh = true
   }
+  $apt_update_successful = @("BASH")
+    for try in {${apt::_update['tries']}..1}; do
+      timeout ${apt::_update['timeout'] ${apt::provider} update && exit 0
```
Comment (kenyon) [inline manifests/update.pp:61]:
```suggestion
      timeout ${apt::_update['timeout']} ${apt::provider} update && exit 0
```

**Review by kenyon — CHANGES_REQUESTED** [review]

File: manifests/update.pp (hunk 1 above)
Comment (pillarsdotnet) [reply manifests/update.pp]: Yeah, that's a bad typo.
Also,  I forgot that Debian doesn't typically install bash as the default shell.
Thirdly, I forgot to escape my dollar-signs.

Fixed all three problems with a re-push.

@kenyon -- Please have another look.

File: manifests/update.pp [hunk 2]
```diff
@@ -56,13 +56,20 @@
...
+  $apt_update_successful = @("ASH"/$)
+    for try in $(seq ${apt::_update['tries']} -1 1); do
+      timeout ${apt::_update['timeout']} ${apt::provider} update && exit 0
+      [ \$try -eq 1 ] && exit \$?
+      sleep 1
+    done
+    - ASH
   exec { 'apt_update':
-    command     => "${apt::provider} update",
+    command     => "${apt::provider} update FAILED",
```
Comment (kenyon) [inline manifests/update.pp:67]: Maybe this should run something more descriptive? When running `apt-get update FAILED` the output is `E: The update command takes no arguments`.

```suggestion
    command     => "/bin/false ${apt::provider} update FAILED",
```

**Comment by pillarsdotnet** [comment]
Thanks, @kenyon -- Fixed along with another trivial error.

**Review by kenyon — DISMISSED** [review]

**Comment by chelnak** [comment]
Hey @pillarsdotnet - appreciate this contribution.

However, I wonder if the changes introduce are slightly hard to follow.

Is there a more idiomatic approach we could take?

**Comment by pillarsdotnet** [comment]
I'm open to suggestions, but this is how I've been resolving the problem locally while this bug stood unresolved for the last two years.

**Comment by LukasAud** [comment]
Hi @pillarsdotnet, it seems like this change is producing a few failures in our automated testing. These need to be addressed before we can think about merging your PR.

**Comment by pillarsdotnet** [comment]
@LukasAud 

I believe the automated testing errors are unrelated to my PR.

To confirm, I have opened #1077

**Review by kenyon — CHANGES_REQUESTED** [review]
Test failures are certainly due to these changes. Maybe need to set the provider to `shell`? Not sure why it's giving an empty string as the command in the error.

**Comment by pillarsdotnet** [comment]
The weird thing is that it's giving an empty string as the command in errors for *other* execs besides the one I modified.

**Comment by pillarsdotnet** [comment]
Okay, setting `provider => shell` helped, but there is also a spec test that needed to be modified.

**Comment by pillarsdotnet** [comment]
Re-coded in case the testing box lacks the `seq` binary.

File: manifests/update.pp [hunk 3]
```diff
@@ -56,13 +56,25 @@
   } else {
     $_refresh = true
   }
+  $apt_update_successful = @("ASH"/$)
+    try=1
+    while [ $try -le ${apt::_update['tries']} ]; do
+      timeout ${apt::_update['timeout']} ${apt::provider} update && exit 0
+      [ \$try -eq 1 ] && exit \$?
```
Comment (kenyon) [inline manifests/update.pp:63]: if you put this into https://www.shellcheck.net/

```sh
#!/bin/sh
    try=1
    while [ $try -le 10 ]; do
      timeout 11 /usr/bin/apt update && exit 0
      [ $try -eq 1 ] && exit $?
      sleep 1
      try=$(( try + 1 ))
    done
```

it finds https://www.shellcheck.net/wiki/SC2319 which I think is a valid mistake here.

**Review by kenyon — CHANGES_REQUESTED** [review]

File: manifests/update.pp (hunk 3 above)
Comment (pillarsdotnet) [reply manifests/update.pp]: @kenyon 
You're probably right. Re-coded; shellcheck is happy now.

Comment (smortex) [reply manifests/update.pp]: I do not understand what this code try to do… Try at most 10 times, exit immediately with 0 when apt succeed and exit with the last try exit code if it failed all the times? Then:

```sh
    try=1
    res=-1
    while [ $res -ne 0 -a $try -le 10 ]; do
      timeout 11 /usr/bin/apt update
      res=$?
      if [ $res -ne 0 ]; then
        sleep 1
        try=$(( try + 1 ))
      fi
    done
    exit $res
```

And what's the point of the timeout (the `exec` resource already has one)?

File: manifests/update.pp [hunk 4]
```diff
@@ -56,13 +56,27 @@
...
     $_refresh = true
   }
+  $apt_update_successful = @("ASH"/$)
+    try=1
+    while [ \$try -le ${apt::_update['tries']} ]; do
+      timeout ${apt::_update['timeout']} ${apt::provider} update
+      code=\$?
+      [ \$code -eq 0 ] && exit 0
+      [ \$try -eq 1 ] && exit \$code
+      sleep 1
```
Comment (smortex) [inline manifests/update.pp:66]: never reached: if `timeout` succeeds, we exit with `0` 2 lines above, otherwise we exit with `$code` 1 line above.

Comment (pillarsdotnet) [reply manifests/update.pp:66]: Line 66 is only reached if $try is greater than 1.

Comment (pillarsdotnet) [reply manifests/update.pp:66]: @smortex

Comment (pillarsdotnet) [reply manifests/update.pp:66]: (looking again)... Crap; you're right. I'm counting up, not down. Gotta compare it to `${apt::update['tries']}`, not to `1`.

Fixed and re-pushed.

File: manifests/update.pp (hunk 3 above)
Comment (pillarsdotnet) [reply manifests/update.pp]: The point of the timeout is that it's a timeout *per try*, not a cumulative total timeout.
I really should remove the timeout from the exec resource but a bunch of rspec tests explicitly look for it and it's easier to leave it in.

Comment (pillarsdotnet) [reply manifests/update.pp]: The point of the entire code is to do everything that the original exec did, only in the `unless` param, instead of in the main exec.

If the `unless` code succeeds, no need to run the main `exec` (and therefore, no corrective change gets logged).

If the `unless` code fails, we want to explicitly log a failure; that is the only job of the main exec.

**Comment by pillarsdotnet** [comment]
I don't understand why a change to `apt_update` would prevent `apt_backports` from working, unless there is a hidden dependency problem.   Perhaps we need to run `apt update` *after* adding backports to sources?

**Comment by pillarsdotnet** [comment]
Maybe I need to run that scriptlet under `sh -c` even though `provider => shell` ?

**Comment by pillarsdotnet** [comment]
Okay; I'm fresh out of ideas. Help, anyone?

**Comment by smortex** [comment]
This PR raise a few concerns for me:

1. Now, each run of puppet runs `apt update` which consume resources on the Debian project side. By default this is every 30mn, so 48 times per day, way more than the default configuration of e.g. `unattended-upgrades` which does it only twice a day.  I do not know if there is some guidelines about how often it is okay to look for updates, but this feels very aggressive;
2. The current behavior seems perfectly valid: running `apt update` is a corrective change (from a state where we do not know *if the repository is up-to-date* to a state where we know that the repository *is currently up-to-date*). So it makes sense that a configuration that enforce "we have up-to-date repositories" appears as having corrective changes all the time.  A way to have a similar behavior without corrective changes would be to use puppet to setup e.g. `unattended-upgrades` to keep the repositories up-to-date and not doing so using puppet.

**Comment by kenyon** [comment]
But, also, the `apt_update` exec usually has `refreshonly => false`, unless `frequency => always`, and `unless` commands only get run if the exec is refreshed.

**Comment by pillarsdotnet** [comment]
@kenyon 

I could try to interpret the output of `apt update` and return nonzero if it actually changes something, then make the main exec something like `echo Repository cache has been updated.`

**Comment by kenyon** [comment]
I don't think `apt update` gives output that would allow that interpretation.

**Comment by pillarsdotnet** [comment]
@kenyon No, but the contents of `/var/cache/apt/pkgcache.bin` changes if and only if `apt update` did something.

**Comment by pillarsdotnet** [comment]
Okay, so tests pass if I explicitly add a `apt update` into the backports acceptance test.  That means the `unless` param to `Exec['apt_update']` isn't doing its job. Maybe I got the syntax wrong for the DASH shell.

(testing by copying the contents into a `test.sh` file and running it under `dash`)

Damnit; had a typo.  Fixing...

**Comment by pillarsdotnet** [comment]
Nope; that didn't fix. I need to see what that scriptlet is doing. Copied it into the failing test, hoping that it will at least output an error message that I can interpret.

**Comment by pillarsdotnet** [comment]
Okay, I'm stumped again.  Help, anyone?

**Comment by pillarsdotnet** [comment]
Modified code to explicitly run the update scriptlet under `/bin/sh -x` and output the results.  Need to see what is going wrong on Debian 10/11. Can someone please approve tests?

**Comment by LTangaF** [comment]
I like what you're doing here. I really wish we could get this moving. It's been years.

**Comment by david22swan** [comment]
Just closing and reopening to re-kick the tests

**Comment by pillarsdotnet** [comment]
Rebased on upstream and added some reasonable defaults for `$apt::_update['tries']` and `$apt::_update['timeout']`.

**Comment by david22swan** [comment]
Closing and reopening to kick tests

File: templates/update_had_no_effect.sh.epp [hunk 5]
```diff
@@ -0,0 +1,21 @@
+set -x
+eval "$(apt-config shell DIR Dir::Cache)"
+cd "/$DIR" || exit 0
+OLD="$(mktemp -p "/$DIR")"
+eval "$(apt-config shell CUR Dir::Cache::pkgcache)"
+ln -f "$CUR" "$OLD"
+TRIES=<%= $apt::_update.fetch('tries', 1) %>
+while true; do
+  if timeout 1 true; then
+    timeout <%= $apt::_update.fetch('timeout', 30)m %> <%= $apt::provider %> update && break
```
Comment (smortex) [inline templates/update_had_no_effect.sh.epp:10]: Typo, as reported by CI:

```suggestion
    timeout <%= $apt::_update.fetch('timeout', 30) %>m <%= $apt::provider %> update && break
```

File: templates/update_had_no_effect.sh.epp [hunk 6]
```diff
@@ -0,0 +1,21 @@
+set -x
```
[the remark this answers is not in the document]
Comment (smortex) [reply templates/update_had_no_effect.sh.epp:1]: This actually is not a shell script, but passed to a shell by the `exec` resource through the `unless` parameter, so a shebang would be ignored and is not required in this case.  Yet, this construct is quite unusual and unexpected, I think it is easy to git tricked into thinking it is a shell script.

File: spec/acceptance/apt_backports_spec.rb [hunk 7]
```diff
@@ -3,6 +3,21 @@
 require 'spec_helper_acceptance'
 
 describe 'apt::backports' do
+  context 'update_had_no_effect script' do
+    let(:pp) do
+      <<-MANIFEST
+        exec { 'Test the script':
+          command   => epp('apt/update_had_no_effect.sh.epp'),
```
Comment (smortex) [inline spec/acceptance/apt_backports_spec.rb:10]: Since this is code in the shell language, I think we need to set `provider => 'shell'` for this test to pass.

**Comment by pillarsdotnet** [comment]
I haven't had time to figure out why the tests are failing.

**Comment by pillarsdotnet** [comment]
Certainly willing to accept help or to drop this approach in favor of a better one.

File: templates/update_had_no_effect.sh.epp (hunk 5 above)
Comment (pillarsdotnet) [reply templates/update_had_no_effect.sh.epp]: Fixed.

File: spec/acceptance/apt_backports_spec.rb (hunk 7 above)
Comment (pillarsdotnet) [reply spec/acceptance/apt_backports_spec.rb:10]: Done.

**Comment by pillarsdotnet** [comment]
@Geod24 @smortex @bastelfreak @david22swan 

So sorry I've let this languish for so long. I'm not allowed to commit to public projects from my work computer and I haven't checked from my PC in a long time.

Please re-kick tests?

**Comment by pillarsdotnet** [comment]
Another test run please, @bastelfreak ?

**Comment by pillarsdotnet** [comment]
Another test run, please, @bastelfreak ?

**Comment by Geod24** [comment]
BTW, to avoid this problem, you should raise a PR against main that fixes something tiny. Once you are recognized as a committer there won't be a need to approve your work every time.

**Comment by pillarsdotnet** [comment]
@bastelfreak -- I think the lint-error is a false-positive, but I've rearranged code to make it happy. Another test, please?

**Comment by pillarsdotnet** [comment]
Well, at least the lint checks pass now.

Made some adjustments that should fix at least the first acceptance test error.

@bastelfreak ?

**Comment by pillarsdotnet** [comment]
Debian-10 is failing on the `eval` line. I've re-coded to use `source` instead of `eval`.

@bastelfreak ?

**Comment by pillarsdotnet** [comment]
Missed one; fixed now.

@bastelfreak ?

**Comment by pillarsdotnet** [comment]
I guess it's possible that `mktemp` and `cmp` are missing in the acceptance-test environment; coding to workaround that possibility.

Another test run, @bastelfreak ?

**Comment by pillarsdotnet** [comment]
Another workaround.

@bastelfreak ?

**Comment by pillarsdotnet** [comment]
Another workaround. @bastelfreak ?

**Comment by pillarsdotnet** [comment]
Yet another workaround. It's frustrating that the "acceptance" environment doesn't actually have any working commands.
@bastelfreak ?

Thanks so much for your patience on this.

**Comment by Geod24** [comment]
@pillarsdotnet : I reiterate my suggestion that you made a minor tweak to main in another PR so that you don't have to ping @bastelfreak for every time you want to run the CI.

**Comment by bastelfreak** [comment]
I also get an email for each push, you don't need to mention me. That will send an additional email :D (also I don't work for puppet and would appreciate the modules team taking care of this).

**Comment by pillarsdotnet** [comment]
I'm not sure why `idempotently_apply()` is failing, but I changed the spec test to better match the code in `manifests/update.pp`

@bastelfreak ?

**Comment by pillarsdotnet** [comment]
> I also get an email for each push, you don't need to mention me. That will send an additional email :D (also I don't work for puppet and would appreciate the modules team taking care of this).

Sorry; I don't know how to ping the modules team. Should I post in slack every time I push a commit?

**Comment by bastelfreak** [comment]
the tests will run when someone with merge permissions presses the button :)
that's a restriction for every new contributor to the repo.

**Comment by pillarsdotnet** [comment]
WOO-HOO!
 
 All acceptance tests that are in any way related to the changes in this PR are now passing.

I've squashed all the commits in order to reduce noise.

Can I get a review, please?

**Comment by pillarsdotnet** [comment]
Some of the other spec tests check whether the `apt::update` exec accepts `timeout` and `tries` params; since it's now a simple `echo`, those  tests are no longer applicable.

**Comment by pillarsdotnet** [comment]
On re-reading the docs for the `exec` resource, it seems that the `timeout` and `tries` params are honored for the `unless` and `onlyif` stanzas as well, so I reverted their removal and accompanying test changes.

**Comment by pillarsdotnet** [comment]
@kenyon ?
@smortex ?

**Review by kenyon — CHANGES_REQUESTED** [review]
I don't really like this because it makes the `apt_update` exec much more complicated. As mentioned in previous comments https://github.com/puppetlabs/puppetlabs-apt/pull/1073#issuecomment-1358569615, the right way to set up a periodic `apt update` is with unattended-upgrades. So then the only time this exec command gets run is when there really is a corrective change, because it has been refreshed by another resource.

File: manifests/update.pp [hunk 8]
```diff
@@ -56,13 +56,25 @@
...
   }
+  # We perform the update in an `unless` clause of the exec, and
+  # return true only if the package cache file changed.
+  # This ensures that Puppet does not report a change if the
+  # update command had no effect. See MODULES-10763 for discussion.
+  $apt_update_had_no_effect = epp(
+    'apt/update_had_no_effect.sh.epp',
+    'provider' => $apt::provider,
+    'timeout'  => $apt::_update['timeout'],
+    'tries'    => $apt::_update['tries'],
```
[the remark this answers is not in the document]
Comment (pillarsdotnet) [reply manifests/update.pp:67]: Sure; Ruby converts trailing named arguments to a hash automatically, but I guess the Puppet DSL makes no such promise.

File: manifests/update.pp [hunk 9]
```diff
@@ -56,13 +56,25 @@
   } else {
     $_refresh = true
   }
+  # We perform the update in an `unless` clause of the exec, and
+  # return true only if the package cache file changed.
+  # This ensures that Puppet does not report a change if the
+  # update command had no effect. See MODULES-10763 for discussion.
+  $apt_update_had_no_effect = epp(
+    'apt/update_had_no_effect.sh.epp',
```
[the remark this answers is not in the document]
Comment (pillarsdotnet) [reply manifests/update.pp:64]: Good catch.

**Comment by pillarsdotnet** [comment]
Thanks, @kenyon 

I've implemented your suggestions.

**Review by bastelfreak — APPROVED** [review]
looks good to me, but I'm not an apt expert.

**Comment by kenyon** [comment]
@pillarsdotnet I don't think that's true about the `yum`/`dnf` package provider. `dnf` works differently from `apt`: it automatically does the equivalent of `apt update` when metadata is out of date. So there is no need to explicitly update the cache, and it's definitely not done on every puppet run.

**Comment by kenyon** [comment]
Also, this PR is for the `apt` module. Package providers are here: https://github.com/puppetlabs/puppet/tree/main/lib/puppet/provider/package

File: manifests/update.pp [hunk 10]
```diff
@@ -56,13 +56,27 @@
...
+    "${module_name}/update_had_no_effect.sh.epp",
+    {
+      'provider' => $apt::provider,
+      'timeout'  => $apt::_update['timeout'],
+      'tries'    => $apt::_update['tries'],
+    }
+  )
   exec { 'apt_update':
-    command     => "${apt::provider} update",
+    command     => "echo ${apt::provider} successfully updated the package cache.",
```
Comment (Ramesh7) [inline manifests/update.pp:72]: command => "echo '${apt::provider} successfully updated the package cache.'",

Comment (Ramesh7) [inline manifests/update.pp:72]: I think we wanted to execute the command here but we are prefixing with echo which means we do nothing except echo which is conflicting with original design.
it should as it is.
```
command => "${apt::provider} update",
```
This will get translate to `apt update` or `dnf update` based on the type of provider.

Comment (pillarsdotnet) [reply manifests/update.pp:72]: @Ramesh7 

You are missing the point.  We want the main exec to be an echo reporting that the package cache was updated.

Actually, a `notice` might be more appropriate.

It's the `unless` clause down below that does the actual work.

**Comment by smortex** [comment]
Hey :wave:!  I don't want to be the guy blocking things but this change does not look correct from my window :disappointed::
1. My base concerns still apply: for me the current behavior is more correct than the proposed one (ref https://github.com/puppetlabs/puppetlabs-apt/pull/1073#issuecomment-1358569615: point 1 is I think no more (`refreshonly` used to be removed IRRC), but point 2 still apply):
     > 2\. The current behavior seems perfectly valid: running `apt update` is a corrective change (from a state where we do not know _if the repository is up-to-date_ to a state where we know that the repository _is currently up-to-date_). So it makes sense that a configuration that enforce "we have up-to-date repositories" appears as having corrective changes all the time.
2. The new implementation look complicated (I think this is mostly due to the way we have 50 lines of code as a single command: writing this as a regular shell script, with vertical space to separate sections, real comments (without epp burden), etc. can improve this);
3. The new behavior seems fragile, for example it seems to return 0 (so no change reported to the user and no issue reported by Puppet) when apt configuration is completely broken, for example if a module setup a non-functional repo (call sequence: `apt-get update` fail and return `1` on each run; maximum attempts reached; EXIT_CODE is set to 1 (good); cache file has not changed on disk; EXIT_CODE is set to 0 (bad)).

**Comment by kenyon** [comment]
@pillarsdotnet can you elaborate on why enabling the apt-daily systemd service and timer that come with apt are not possible for you? There is no "re-tooling" to be done, you already have it because it comes with apt. That is the standard Debian way of keeping the package cache up to date: https://debian-handbook.info/browse/stable/sect.regular-upgrades.html (which the [unattended_upgrades module](https://github.com/voxpupuli/puppet-unattended_upgrades) can configure, and it is not required to also enable automatic package upgrades)

If you wanted to run `apt-daily.service` every 30 minutes like what puppet would do with this PR, you would just use a systemd drop-in (using the [systemd module](https://github.com/voxpupuli/puppet-systemd)) to edit the `[Timer]` section of `/lib/systemd/system/apt-daily-upgrade.timer`.

**Comment by skyamgarp** [comment]
Closing due to inactivity, please feel free to reopen when ready. Thanks.
